lbry/wallet/server/db/trending/variable_decay.py

```diff
@@ -60,185 +59,331 @@ def check_trending_values(connection):
         fout = open("trending_variable_decay.log", "a")
         fout.write(s)
         fout.flush()
         fout.close()
 
-class TrendingData:
+
+def trending_unit(height):
     """
-    An object of this class holds trending data
+    Return the trending score unit at a given height.
     """
-    def __init__(self):
+    # Round to the beginning of a SAVE_INTERVAL batch of blocks.
+    _height = height - (height % SAVE_INTERVAL)
+    return 1.0/DECAY**(height % RENORM_INTERVAL)
 
-        # Dict from claim id to some trending info.
-        # Units are TIME VARIABLE in here
-        self.claims = {}
 
-        # Claims with >= WHALE_THRESHOLD LBC total amount
-        self.whales = set([])
+class TrendingDB:
+    """
+    An in-memory database of trending scores
+    """
 
-        # Have all claims been read from db yet?
+    def __init__(self):
+        self.conn = apsw.Connection(":memory:")
+        self.cursor = self.conn.cursor()
         self.initialised = False
+        self.write_needed = set()
 
-        # List of pending spikes.
-        # Units are CONSTANT in here
-        self.pending_spikes = []
+    def execute(self, query, *args, **kwargs):
+        return self.cursor.execute(query, *args, **kwargs)
 
-    def insert_claim_from_load(self, height, claim_hash, trending_score, total_amount):
-        assert not self.initialised
-        self.claims[claim_hash] = {"trending_score": trending_score,
-                                   "total_amount": total_amount,
-                                   "changed": False}
+    def executemany(self, query, *args, **kwargs):
+        return self.cursor.executemany(query, *args, **kwargs)
 
-        if trending_score >= WHALE_THRESHOLD*get_time_boost(height):
-            self.add_whale(claim_hash)
+    def begin(self):
+        self.execute("BEGIN;")
 
-    def add_whale(self, claim_hash):
-        self.whales.add(claim_hash)
+    def commit(self):
+        self.execute("COMMIT;")
+
+    def initialise(self, db):
+        """
+        Pass in claims.db
+        """
+        if self.initialised:
+            return
+
+        trending_log("Initialising trending database...")
+
+        # The need for speed
+        self.execute("PRAGMA JOURNAL_MODE=OFF;")
+        self.execute("PRAGMA SYNCHRONOUS=0;")
+
+        self.begin()
+
+        # Create the tables
+        self.execute("""
+            CREATE TABLE IF NOT EXISTS claims
+                (claim_hash     BYTES PRIMARY KEY,
+                 lbc            REAL NOT NULL DEFAULT 0.0,
+                 trending_score REAL NOT NULL DEFAULT 0.0)
+            WITHOUT ROWID;""")
+
+        self.execute("""
+            CREATE TABLE IF NOT EXISTS spikes
+                (id         INTEGER PRIMARY KEY,
+                 claim_hash BYTES NOT NULL,
+                 height     INTEGER NOT NULL,
+                 mass       REAL NOT NULL,
+                 FOREIGN KEY (claim_hash)
+                    REFERENCES claims (claim_hash));""")
```

Comment:
foreign keys and references add significant performance penalty

Reply:
I didn't know that. I guess for speed I could leave them implicit. Do you want me to make these changes? It is unclear.

Reply:
for sqlite it might not be a problem, i don't think we have FK enforcement enabled anyways, but it's something to keep in mind when we add support for postgresql where FKs get checked on every insert/update and has noticeable performance impact when doing bulk inserts / updates